### ISSUE TITLE
Additional fieldSolverNone necessary with EMF centered in cell?

### ISSUE THREAD

**Issue by n01r** [ComputationalRadiationPhysics/picongpu#1072]
Apart from the fact that field solvers should probably also be `struct`s, just as pushers and ionizers are now:   
*Should there be two (more?) specializations of the `fieldSolverNone` for each `NumericalCellType`?*

Currently this one just has the Yee cell:
```
namespace fieldSolverNone
{
    typedef picongpu::noSolver::NoSolver FieldSolver;
    typedef yeeCell::YeeCell NumericalCellType;
}
```
It might be useful to have the electromagnetic fields centered in the cell when they are defined by the background field and not the usual laser definition.   
I'd propose something like:
```
namespace picongpu
{
namespace fields
{
namespace MaxwellSolver
{
template<T_NumericalCellType>
struct None
{
    typedef fields::numericalCellTypes::T_NumericalCellType NumericalCellType
}
} // namespace MaxwellSolver
} // namespace fields
} // namespace picongpu
```

**Comment by ax3l** [comment]
let's keep thing simple.
it's very convenient that field solvers determine the cell type.

nevertheless, I know some field solvers that could be in principle be implemented for centered and staggered fields.

Using a templated parameter for those (and `fieldSolverNone`) might be a good idea, for the others we could leave the cellType selection in the unitless file.

introducing the celltype in all solvers as an argument might be more consistent, we should evaluate what causes more/less confusion.

**Comment by ax3l** [comment]
for the meantime: simply add the `unitless` file to your case and change it there. that is exactly the use case for it (advanced option that should not (yet) be exposed to non power-uses).

**Comment by ax3l** [comment]
also the documentation is not 100% correct
```
fieldSolverNone: disable the vacuum update of E and B
```

the field solver also determines how `J` influences the change of `E` (and `B`).
`fieldSolverNone` only disables

**Comment by n01r** [comment]
Also:   
`src/picongpu/include/fields/MaxwellSolver/None/NoSolver.hpp`:

```
template<typename T_Dummy>
struct ConditionCheck<NoSolver, T_Dummy>
{
    /* Courant-Friedrichs-Levy-Condition for Yee Field Solver: */
    PMACC_CASSERT_MSG(Courant_Friedrichs_Levy_condition_failure____check_your_gridConfig_param_file,
        (SPEED_OF_LIGHT*SPEED_OF_LIGHT*DELTA_T*DELTA_T*INV_CELL2_SUM)<=1.0);
};
```

IMO when `fieldSolverNone` is chosen the compile process should not fail because of a violation of the CFL criterion. Instead a warning should be thrown both while compiling and later in the simulation output.

What do you guys think, @ax3l , @psychocoderHPC ?

**Comment by ax3l** [comment]
there is still a criterion close to "CFL" which is: a particle shall not
travel more than a cell per timestep.

it's a bit more relaxed (min(cell_edge) / c) but still needs to be true
as long as particles are allowed to move.

this is an internal limitation of libPMacc for moving particles.


but you are right, for Pusher == None and Solver == None we can remove it.

**Comment by n01r** [comment]
Yeah, at least if both Pusher and Solver == None. But it's important that there is still a warning.

**Comment by ax3l** [comment]
> But it's important that there is still a warning.

why should this be important?

**Comment by ax3l** [comment]
I think I will do the following: no cassert in solver none and an additional assert in each pusher (but pusher none) for `min(cellEdge) / c < dt`

**Comment by n01r** [comment]
Let's say people get the idea to test sth first with a background field an no active pushers and field solvers. Then they want to make it more complicated by allowing particles to move. They should know then what they're doing wrong and how to correct it. They should also be aware of it before starting the simulation. Therefore the warning ;)

**Comment by ax3l** [comment]
I disagree, warnings that are unnecessary just create noise and distract users from reading warnings that are relevant.

then they get an error in case two. a warning in case one is not useful, as it is absolutely irrelevant as in your test case.

**Comment by n01r** [comment]
alright, convinced

**Comment by ax3l** [comment]
sweet, we'll update the casserts soon - so long just comment them out pls to continue with your no-pusher & no-solver test :)

**Comment by ax3l** [comment]
P.S.: next time, pls do open a new issue for new topics (https://github.com/ComputationalRadiationPhysics/picongpu/issues/1089)
